pkg/alibabacloud/eni/instances.go

```diff
@@ -0,0 +1,240 @@
+		return time.Time{}
+	}
+
+	instances, err := m.api.GetInstances(ctx, vpcs, vSwitches)
+	if err != nil {
+		log.WithError(err).Warning("Unable to synchronize ECS interface list")
+		return time.Time{}
+	}
+
+	log.WithFields(logrus.Fields{
+		"numInstances":      instances.NumInstances(),
+		"numVPCs":           len(vpcs),
+		"numVSwitches":      len(vSwitches),
+		"numSecurityGroups": len(securityGroups),
+	}).Info("Synchronized ENI information")
+
+	m.mutex.Lock()
+	m.instances = instances
+	m.vSwitches = vSwitches
+	m.vpcs = vpcs
+	m.securityGroups = securityGroups
+	m.mutex.Unlock()
+
+	return resyncStart
+}
+
+// GetVSwitches returns all the tracked vSwitches
+// The returned subnetMap is immutable so it can be safely accessed
+func (m *InstancesManager) GetVSwitches() ipamTypes.SubnetMap {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	subnetsCopy := make(ipamTypes.SubnetMap)
+	for k, v := range m.vSwitches {
+		subnetsCopy[k] = v
+	}
+
+	return subnetsCopy
+}
+
+// GetVSwitch return vSwitch by id
+func (m *InstancesManager) GetVSwitch(id string) *ipamTypes.Subnet {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+	return m.vSwitches[id]
+}
+
+// ForeachInstance will iterate over each instance inside `instances`, and call
+// `fn`. This function is read-locked for the entire execution.
+func (m *InstancesManager) ForeachInstance(instanceID string, fn ipamTypes.InterfaceIterator) {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+	m.instances.ForeachInterface(instanceID, fn)
+}
+
+// UpdateENI updates the ENI definition of an ENI for a particular instance. If
+// the ENI is already known, the definition is updated, otherwise the ENI is
+// added to the instance.
+func (m *InstancesManager) UpdateENI(instanceID string, eni *eniTypes.ENI) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	eniRevision := ipamTypes.InterfaceRevision{Resource: eni}
+	m.instances.Update(instanceID, eniRevision)
+}
+
+// AllocIndexForENI will alloc an monotonically increased index for each ENI on this instance.
+// The index generated the first time this ENI is created, and stored in ENI.Tags.
+func (m *InstancesManager) AllocIndexForENI(instanceID string) (int, error) {
+	// alloc index for each created ENI
+	used := make([]bool, maxENIPerNode)
+	err := m.instances.ForeachInterface(instanceID, func(instanceID, interfaceID string, iface ipamTypes.InterfaceRevision) error {
+		cachedENI, ok := iface.Resource.(*eniTypes.ENI)
+		if !ok {
+			return fmt.Errorf("ENI type mismatch, got %T", iface.Resource)
+		}
+		index := utils.GetENIIndexFromTags(cachedENI.Tags)
+		if index > maxENIPerNode || index < 0 {
+			return fmt.Errorf("ENI index(%d) is out of range", index)
+		}
+		used[index] = true
+		return nil
+	})
+	if err != nil {
+		log.WithError(err).Warning("Unable to retrieve index for ENI")
+		return 0, err
+	}
+	// ECS has at least 1 ENi, 0 is reserved for eth0
+	i := 1
+	for ; i < maxENIPerNode; i++ {
+		if !used[i] {
+			break
+		}
+	}
```

Comment:
Nice solution for monotonically increasing index :). It would be nice to add unit tests for this just to validate that we haven't made any mistakes